### ISSUE TITLE
Example FMU Errors

### ISSUE THREAD

**Issue by rukie** [NTNU-IHB/PythonFMU#112]
I've taken your example and tried to import into a modelica based tool.
I think it would be good to have an example FMU that runs without errors..
```
Model description not ok. There are some error(s).
Error(s):
Scalar variable intOut cannot have continuous variability since it is not a real variable
Scalar variable booleanVariable cannot have continuous variability since it is not a real variable
Scalar variable stringVariable cannot have continuous variability since it is not a real variable
Warning(s):
Missing InitialUnknown index 1, 2
String-type scalar variable stringVariable is ignored by Amesim. It will be initialized to its start value.
There are some errors.
```

I am still learning about this process, so I don't have suggestions on a fix yet. I'll share what I learn as I get there though!

**Comment by rukie** [comment]
At the very least, I removed the intOut, booleanVariable, and stringVariable.
This left the realOut variable. 

Upon Amesim trying to compile the FMU, it complains CSIMPPythonSlave.obj does not exist. I do however, get a CSIMPPythonSlave.o file.

**Comment by markaren** [comment]
> I think it would be good to have an example FMU that runs without errors..

These are trival "errors". The reason why they are `continuous` by defualt is becouse of a minor imiguity in the standard

`
"continuous": Only a variable of type = “Real” can be “continuous”.
ModelExchange: No restrictions on value changes.
Co-Simulation: By convention, the variable is from a differential
The default is “continuous”.
`

While it says that only reals can be `continuous`, it also says that it is the default (which implies for any variables).
Failing on this is overly strict from Amesim, however a solution could be to not specify a variability by default. 

> Missing InitialUnknown index 1, 2

This is just a warning. Not sure if or when I will fix it.

> Upon Amesim trying to compile the FMU

Are you talking about a source code FMU here? I'll doubt that will work with PythonFMU.

**Comment by markaren** [comment]
Actually, I just checked. PythonFMU does not declare continuous variability by default. So it sounds like an import error, not PythonFMU error. Your tool probably translates no variability to continuous when importing the XML ?

**Comment by rukie** [comment]
AMESIM is a tool like OpenModelica, but so how it imports the FMU is a little bit of a mystery. I regularly work with MSC Adams and Siemen's Simcenter Amesim. I'm investigating both your tool and https://github.com/ghorwin/FMICodeGenerator as a means of using text based code for system/vehicle controls rather than Mathworks Simulink. 

Generally all the variables of interest are real floats. My experience so far with the tools is that they take the FMU source files and compile on import. Amesim utilizes the gcc compiler by default (I can run on Windows or Linux). 

I'm also investigating a C code style FMU system for embedded controllers, looking at https://github.com/ghorwin/FMICodeGenerator heavily. 

I'm more of a python scripter than I am a developer of anything, so please pardon my ignorance as I get up to speed :)

After going through the C tool, I understand the initial value should be set within the FMI/FMU code but can work without it. 

**The errors in the first post are generally resolved, but now Amesim is looking for a *.obj file when CMAKE only releases a *.o file.**

**Comment by markaren** [comment]
The reason I say that source code FMU probably won't work with PythonFMU is that, since it needs to link with Python3 in order to compile, having just the source files from the FMU (C/C++ headers and source files) is not enough. How would the build system know that it should link with Python3.dll ? With that being said I never looked that much into how the standard describes the use of source code FMUs.

**Comment by rukie** [comment]
Can you elaborate on source-code FMU? I'm familiar with the points model exchange versus co-simulation, but I'm getting a little lost with the source code reference. As I understand the "source code" is optional, with binary files being supported. If the software is FMI compliant that would be sufficient?

**Comment by markaren** [comment]
You are saying that AMESIM _compiles_ the FMU on import. This would mean that it tries to import the FMU as a source code FMU. In a normal FMI import context, the tool would just import and run the FMU since the binaries are already pre-compiled.

**Comment by markaren** [comment]
You are not trying to build an FMU from the master branch are you? No pre-compiled binaries are available in this case, perhaps AMESIM thinks that since there are no binaries it should generate them itself. 
PythonFMU is meant to be installed using pip/conda.

**Comment by rukie** [comment]
I used conda to install pythonfmu. 

On import I get the following message:
```
>Starting: "C:\Program Files\Simcenter\2020.1\Amesim\win64\fmi_import_compiler.exe -outdir C:/MS/Projects/WR5609_Embedded_Controller_CoSim/AME_WD/pysubmodel -ccopt -win64_gcc -preferred cs C:/MS/Projects/WR5609_Embedded_Controller_CoSim/pyFMU/PythonSlave.fmu"

It is an FMU for Co-simulation, version 2.0, running on linux64, win64
Submodels Windows compilation batch file.
-----------------------------------------------------------------------
This batch file is to automate the compile procedure for submodels 
or other userwritten C files using the GCC 64x compiler. 
-----------------------------------------------------------------------
- Will compile the following file(s):
CSIMPPythonSlave.c        

C:\MS\Projects\WR5609_Embedded_Controller_CoSim\AME_WD\pysubmodel\submodels>gcc -DWIN64 -DWIN32 -m64 -O1 -DNDEBUG -D_WINDOWS -D_MBCS -I"C:\Program Files\Simcenter\2020.1\Amesim"\lib -c CSIMPPythonSlave.c -o win64-gcc\CSIMPPythonSlave.o 
Submodel build completed!

Model description ok.
Warning(s):
Missing InitialUnknown index 1
Creating C:/MS/Projects/WR5609_Embedded_Controller_CoSim/AME_WD/pysubmodel/./submodels/CSIMPPythonSlave.spe submodel file...
Creating C:/MS/Projects/WR5609_Embedded_Controller_CoSim/AME_WD/pysubmodel/./submodels/CSIMPPythonSlave.c submodel source file...
```

and that gives some output
```
CSIMPPythonSlave.c:33:0: warning: "_WIN32_WINNT" redefined
 #define _WIN32_WINNT 0x0502 
 ^
In file included from C:/Program Files/Simcenter/2020.1/Amesim/sys/mingw64/x86_64-w64-mingw32/include/crtdefs.h:10:0,
                 from C:/Program Files/Simcenter/2020.1/Amesim/sys/mingw64/x86_64-w64-mingw32/include/math.h:13,
                 from CSIMPPythonSlave.c:7:
C:/Program Files/Simcenter/2020.1/Amesim/sys/mingw64/x86_64-w64-mingw32/include/_mingw.h:225:0: note: this is the location of the previous definition
 #define _WIN32_WINNT 0x502
 ^
CSIMPPythonSlave.c: In function 'GetDLLfullPath':
CSIMPPythonSlave.c:742:21: warning: assignment discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
       actualFmuPath = fmuPath;  /* fallback */
```

at which point I'm able to place the FMU on the sketch. 

When I go to 'compile' the sketch, much like one would in OpenModelica, I get a little error about the FMU.
![image](https://user-images.githubusercontent.com/6559628/107971593-4b004680-6f78-11eb-8229-c4b66e9f0fdd.png)

I will follow up with the software developer to see if there's something off on their end.

**Comment by markaren** [comment]
Just to make sure. You are using pythonfmu 0.6.1 and not pyfmu, which is a similar project right?

**Comment by rukie** [comment]
Yes I am using pythonfmu. Once I get a license to export FMUs from Amesim, I could try importing the FMU into pyFMI. I wasn't aware of pyFMU though which appears deprecated and replaced by [unifmu](https://github.com/INTO-CPS-Association/unifmu).

**Comment by markaren** [comment]
> Once I get a license to export FMUs from Amesim

I gotta ask. Are you using pythonfmu in a way I dont understand? You copied the code from the readme into a file named e.g. demo.py file and ran pythonfmu build -f demo.py right? Then you are trying to import the generated FMU into AMEsim and you get the error you are posting? 

I think er need to clarify what you are trying to do...

**Comment by rukie** [comment]
You posted pretty much what I'm trying to do.

My goal is take a modelica model of a physical system, imagine a hydraulic cylinder, and control its position with text based code. 
I can achieve this by either taking the modelica model and making an FMU and importing into python,
or by making an FMU of the python code/controller and importing into modelica.

In this example Amesim = Modelica.

**Comment by markaren** [comment]
Ok, nothing about the error messages you have posted so far seems related to pythonfmu, so I'd take it up with Amesim.
I have no idea what `CSIMPPythonSlave.c` and the C code from the error is not from the python/C++ bridge that pythonfmu uses.
